### PR TITLE
Gaojing/devicelogin exit when interactivemode never

### PR DESCRIPTION
**Pull request by everjing**
"Printed out "KUBERNETES_EXEC_INFO" in the code for verification purpose and enabled to return error under condition of "interactiveMode : Never" under devicelogin method. Will do another commit once its logic part is approved. Please check it out. Thanks!

### CODE REVIEW

File: pkg/cmd/token.go [hunk 1]
```diff
@@ -3,6 +3,7 @@ package cmd
 import (
 	"github.com/Azure/kubelogin/pkg/token"
```
Comment (weinong) [inline pkg/cmd/token.go:4]: can we revert the change in this file as .... there is no change?

File: pkg/token/execCredentialPlugin.go [hunk 2]
```diff
@@ -28,6 +32,16 @@ type execCredentialPlugin struct {
 }
 
 func New(o *Options) (ExecCredentialPlugin, error) {
+	env := os.Getenv(execInfoEnv)
+	fmt.Fprintln(os.Stderr, os.Getenv(execInfoEnv))
```
Comment (weinong) [inline pkg/token/execCredentialPlugin.go:36]: this line is for debugging, right? you can log it using klog.V(10).Info(), though

File: pkg/token/execCredentialPlugin.go [hunk 3]
```diff
@@ -28,6 +32,16 @@ type execCredentialPlugin struct {
...
+	var execCredential clientauthentication.ExecCredential
+	error := json.Unmarshal([]byte(env), &execCredential)
```
Comment (weinong) [inline pkg/token/execCredentialPlugin.go:38]: a more idiomatic way is
if err := ...; err != nil {
}

File: pkg/token/execCredentialPlugin.go [hunk 4]
```diff
@@ -28,6 +32,16 @@ type execCredentialPlugin struct {
...
+	if error != nil {
+		return nil, fmt.Errorf("cannot convert to ExecCredential: %w", error)
+	}
+	if !execCredential.Spec.Interactive && o.LoginMethod == "DeviceCodeLogin" {
```
Comment (weinong) [inline pkg/token/execCredentialPlugin.go:42]: are you sure it's `"DeviceCodeLogin"` instead of the const?

File: pkg/token/execCredentialPlugin.go (hunk 2 above)
Comment (weinong) [reply pkg/token/execCredentialPlugin.go:36]: besides, you already have `env`. don't need to `os.Getenv()` again?

File: pkg/cmd/token.go (hunk 1 above)
Comment (everjing) [reply pkg/cmd/token.go:4]: Done for next commit.

File: pkg/token/execCredentialPlugin.go (hunk 2 above)
Comment (everjing) [reply pkg/token/execCredentialPlugin.go:36]: Addressed for next commit.

File: pkg/token/execCredentialPlugin.go (hunk 3 above)
Comment (everjing) [reply pkg/token/execCredentialPlugin.go:38]: Changed per comments and please check next commit. Thanks!

File: pkg/token/execCredentialPlugin.go (hunk 4 above)
Comment (everjing) [reply pkg/token/execCredentialPlugin.go:42]: I checked it and you are right. Changed to  DeviceCodeLogin const. Thanks!

File: pkg/cmd/token.go [hunk 5]
```diff
@@ -19,7 +19,6 @@ func NewTokenCmd() *cobra.Command {
 			if err := o.Validate(); err != nil {
 				return err
 			}
-
```
Comment (weinong) [inline pkg/cmd/token.go:22]: i feel this is a unnecessary change?

File: pkg/token/execCredentialPlugin.go [hunk 6]
```diff
@@ -28,6 +32,15 @@ type execCredentialPlugin struct {
 }
 
 func New(o *Options) (ExecCredentialPlugin, error) {
+	env := os.Getenv(execInfoEnv)
+	//fmt.Fprintln(os.Stderr, os.Getenv(execInfoEnv))
```
Comment (weinong) [inline pkg/token/execCredentialPlugin.go:36]: please remove unnecessary commented out code. for debugging purpose, use klog with higher verbosity

File: pkg/token/execCredentialPlugin.go [hunk 7]
```diff
@@ -28,6 +32,15 @@ type execCredentialPlugin struct {
...
+	var execCredential clientauthentication.ExecCredential
+	if error := json.Unmarshal([]byte(env), &execCredential); error != nil {
+		return nil, fmt.Errorf("cannot convert to ExecCredential: %w", error)
+	}
+	if !execCredential.Spec.Interactive && o.LoginMethod == DeviceCodeLogin {
```
Comment (weinong) [inline pkg/token/execCredentialPlugin.go:41]: lets work on #140 first so that you can add appropriate unit test in this PR

File: pkg/cmd/token.go (hunk 5 above)
Comment (everjing) [reply pkg/cmd/token.go:22]: Reverted for next commit.

File: pkg/token/execCredentialPlugin.go (hunk 6 above)
Comment (everjing) [reply pkg/token/execCredentialPlugin.go:36]: Done for next commit. Thanks!

File: pkg/token/execCredentialPlugin.go (hunk 7 above)
Comment (everjing) [reply pkg/token/execCredentialPlugin.go:39]: Added unit test for this commit 8523f02, please check it out. Thanks!

File: pkg/token/execCredentialPlugin.go [hunk 8]
```diff
@@ -29,6 +31,14 @@ type execCredentialPlugin struct {
 }
 
 func New(o *Options) (ExecCredentialPlugin, error) {
+	env := os.Getenv(execInfoEnv)
+	var execCredential clientauthentication.ExecCredential
+	if error := json.Unmarshal([]byte(env), &execCredential); error != nil {
```
Comment (weinong) [inline pkg/token/execCredentialPlugin.go:36]: `error` is a type so generally you don't use the type name as variable name. conventionally, `err` is used

Comment (everjing) [reply pkg/token/execCredentialPlugin.go:36]: Updated with err for next commit. Thanks!

**Review by weinong — APPROVED** [review]
LGTM

File: pkg/token/execCredentialPlugin.go [hunk 9]
```diff
@@ -29,6 +31,14 @@ type execCredentialPlugin struct {
 }
 
 func New(o *Options) (ExecCredentialPlugin, error) {
+	env := os.Getenv(execInfoEnv)
+	var execCredential clientauthentication.ExecCredential
+	if err := json.Unmarshal([]byte(env), &execCredential); err != nil {
```
Comment (yurrriq) [inline pkg/token/execCredentialPlugin.go:36]: Is it now required to specify `KUBERNETES_EXEC_INFO`? My workflow for v0.0.20 now throws `Error: cannot convert to ExecCredential: unexpected end of JSON input`. It seems perhaps because `env` here is `""` which then can't be unmarshalled into a `clientauthentication.ExecCredential`.
